src/game_events/handlers.cpp

```diff
@@ -62,15 +104,263 @@ void event_handler::handle_event(const queued_event& event_info, game_lua_kernel
+};
+
+struct filter_unit : public event_filter {
+	filter_unit(const vconfig& cfg, bool first) : suf_(cfg.make_safe()), first_(first) {}
+	bool operator()(const queued_event& event_info) const override
+	{
+		const auto& loc = first_ ? event_info.loc1 : event_info.loc2;
+		auto unit = resources::gameboard->units().find(loc);
+		return loc.matches_unit_filter(unit, suf_);
+	}
+	void serialize(config& cfg) const override
+	{
+		cfg.add_child(first_ ? "filter" : "filter_second", suf_.to_config());
+	}
+	bool can_serialize() const override
+	{
+		return true;
+	}
+private:
+	unit_filter suf_;
+	bool first_;
+};
+
+struct filter_attack : public event_filter {
+	filter_attack(const vconfig& cfg, bool first) : swf_(cfg.make_safe()), first_(first) {}
+	bool operator()(const queued_event& event_info) const override
+	{
+		const unit_map& units = resources::gameboard->units();
+		const auto& loc = first_ ? event_info.loc1 : event_info.loc2;
+		auto unit = units.find(loc);
+		if(unit != units.end() && loc.matches_unit(unit)) {
+			const config& attack = event_info.data.child(first_ ? "first" : "second");
+			return swf_.empty() || matches_special_filter(attack, swf_);
+		}
+		return false;
+	}
+	void serialize(config& cfg) const override
+	{
+		cfg.add_child(first_ ? "filter_attack" : "filter_second_attack", swf_.get_config());
+	}
+	bool can_serialize() const override
+	{
+		return true;
+	}
+private:
+	vconfig swf_;
+	bool first_;
+};
+
+struct filter_formula : public event_filter {
+	filter_formula(const std::string& formula) : formula_(formula) {}
+	bool operator()(const queued_event& event_info) const override
+	{
+		wfl::gamestate_callable gs;
+		wfl::event_callable evt(event_info);
+		wfl::formula_callable_with_backup data(evt, gs);
+		return formula_.evaluate(data).as_bool();
+	}
+	void serialize(config& cfg) const override
+	{
+		std::string code = formula_.str();
+		if(cfg.has_attribute("filter_formula")) {
+			// This will probably never happen in practice, but handle it just in case it somehow can
+			code = "(" + cfg["filter_formula"].str() + ") and (" + code + ")";
+		}
+		cfg["filter_formula"] = code;
+	}
+	bool can_serialize() const override
+	{
+		return true;
+	}
+private:
+	wfl::formula formula_;
+};
+
+static std::unique_ptr<event_filter> make_filter(const std::string& key, const vconfig& contents)
+{
+	if(key == "filter_condition") {
+		return std::make_unique<filter_condition>(contents);
+	} else if(key == "filter_side") {
+		return std::make_unique<filter_side>(contents);
+	} else if(key == "filter") {
+		return std::make_unique<filter_unit>(contents, true);
+	} else if(key == "filter_attack") {
+		return std::make_unique<filter_attack>(contents, true);
+	} else if(key == "filter_second") {
+		return std::make_unique<filter_unit>(contents, false);
+	} else if(key == "filter_second_attack") {
+		return std::make_unique<filter_attack>(contents, false);
+	}
+	return nullptr;
+}
+
+struct filter_dynamic : public event_filter {
```

Comment:
Please document `filter_dynamic`, what it does isn't obvious from the name.

From the Wiki docs for `[insert_tag]` I'd expect the value of the variable to be evaluated once, at the time that the event is created. If I'm understanding this code correctly, then it gets reevaluated each time the filter runs - I can see usecases that would enable, but it would break WML that constructed the filter with something unit-specific during a `first_time_only=no` event, `[insert_tag]`'d it, and then overwrote it when that `first_time_only=no` event was triggered again by a different unit.

Reply:
That's not how events are evaluated in the engine. The content of an `[event]` tag is divided into two distinct parts: the event definition (including name, ID, and filters), and the ActionWML. Variables in the event definition are substituted when the event triggers, _not_ when the event handler is added.

Your example of something this would break would already break in 1.16.x. Prior to the commit that introduced this class, there was a unit test failure because such dynamic filters no longer worked; this fixes that unit test failure.

Reply:
Still no documentation for this class in the .cpp/.hpp files.

The test was added in #6785, will comment about the Wiki there / in IRC.